src/Tribe/Context.php

```diff
@@ -91,41 +138,260 @@ public function is_editing_post( $post_or_type = null ) {
-	 * context; passing a value to this argument will set it to that value in future checks, a test-related usage.
+	 * context.
 	 *
 	 * @since 4.7.12
-	 *
-	 * @param bool $doing_ajax An injectable status to override the `DOING_AJAX` check.
+	 * @since TBD Removed the $doing_ajax parameter.
 	 *
 	 * @return boolean
 	 */
-	public function doing_ajax( $doing_ajax = null ) {
-		if ( null !== $doing_ajax ) {
-			$this->doing_ajax = (bool) $doing_ajax;
-		} else {
-			$this->doing_ajax = defined( 'DOING_AJAX' ) && DOING_AJAX;
-		}
-
-		return $this->doing_ajax;
+	public function doing_ajax() {
+		return function_exists( 'wp_doing_ajax' )
+			? wp_doing_ajax()
+			: defined( 'DOING_AJAX' ) && DOING_AJAX;
 	}
 
 	/**
 	 * Checks whether the context of the current HTTP request is a Cron one or not.
 	 *
 	 * @since 4.7.23
+	 * @since TBD Removed the $doing_cron parameter.
 	 *
-	 * @param bool|null $doing_cron If set then this method will act as a setter; the current
-	 *                         method call, and the following ones, will return this value.
+	 * @return bool Whether the context of the current HTTP request is a Cron one or not.
+	 */
+	public function doing_cron() {
+		return function_exists( 'wp_doing_cron' )
+			? wp_doing_cron()
+			: defined( 'DOING_CRON' ) && DOING_CRON;
+	}
+
+	/**
+	 * Returns the current display mode.
 	 *
-	 * @return bool whether the context of the current HTTP request is a Cron one or not.
+	 * This is read from the the request variables first and from the main query object second.
+	 *
+	 * @since TBD
+	 *
+	 * @return string The global event display or an empty string if not set in any of the sources above.
 	 */
-	public function doing_cron( $doing_cron = null ) {
-		if ( null !== $doing_cron ) {
-			$this->doing_cron = (bool) $doing_cron;
+	public function get_event_display() {
+		$request_value = tribe_get_request_var( 'tribe_event_display' );
+		global $wp_query;
+		$query_value = $wp_query->get( 'eventDisplay', '' );
+
+		return $request_value ? $request_value : $query_value;
+	}
+
+	/**
+	 * Gets a value reading it from the location(s) defined in the `Tribe__Context::$props
+	 *
+	 * @since TBD
+	 *
+	 * @param string $key The key of the variable to fetch.
+	 * @param mixed|null $default The default value to return if not found.
+	 *
+	 * @return mixed The value from the first location that can provide it or the default
+	 *               value if not found.
+	 */
+	public function get( $key, $default = null ) {
+		$value = $default;
+		$locations = self::$locations[ $key ];
+
+		if ( ! isset( $locations ) ) {
+			return $value;
+		}
+
+		if ( isset( $this->request_cache[ $key ] ) ) {
+			$value = $this->request_cache[ $key ];
 		} else {
-			$this->doing_cron = defined( 'DOING_CRON' ) && DOING_CRON;
+			foreach ( $locations as $location => $keys ) {
+				$keys = (array) $keys;
+
+				switch ( $location ) {
```

Comment:
Would love to move away from `switch` into using:

```
$this->allowed_locations = [ 'request_var', 'query_var', 'tribe_option', 'transient', 'constant', 'global_var', 'static_prop', 'prop', 'static_method', 'method', 'function' ];

return call_user_func_array( [ $this, $location ], [ $keys, $default ] );
```

Reply:
Would you give `public` visibilty to these methods? E.g.:
```php
class Tribe__Context{
	public function request_var($key){
		// ...
	}
}
```

I would say no, exposing these functions would make the `set` method "disappear" in the object API noise and defeat the purpose. I'd start with `protected` visibility to open up if we really find ourselves needing it.

Reply:
Uhnn I see what you mean, maybe we have a second class to deal with the locations. But since we dont have to use PHP 5.2 anymore, we could probably used `$this->$location( $keys, $default )` to a protected method.

I really wish we didn't have such a huge method on the `get`.

Reply:
I've refactored the code to:
1. remove the switch and use functions, `protected` for the time being.
2. refined the fetching to make it so that the lookup will not stop when encountering the first value that's equal to the default one when the default value is `false` or an empty string.